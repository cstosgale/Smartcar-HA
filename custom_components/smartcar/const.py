from enum import StrEnum, auto

DOMAIN = "smartcar"
DEFAULT_NAME = "Smartcar"
API_HOST = "https://api.smartcar.com"

PLATFORMS = [
    "sensor",
    "switch",
    "lock",
    "device_tracker",
    "binary_sensor",
    "number",
]

OAUTH2_AUTHORIZE = "https://connect.smartcar.com/oauth/authorize"
OAUTH2_TOKEN = "https://auth.smartcar.com/oauth/token"
SMARTCAR_MODE = "live"


class Scope(StrEnum):
    READ_VEHICLE_INFO = auto()
    READ_VIN = auto()
    READ_BATTERY = auto()
    READ_CHARGE = auto()
    READ_ENGINE_OIL = auto()
    READ_FUEL = auto()
    READ_LOCATION = auto()
    READ_ODOMETER = auto()
    READ_SECURITY = auto()
    READ_TIRES = auto()
    CONTROL_CHARGE = auto()
    CONTROL_SECURITY = auto()


REQUIRED_SCOPES = [
    Scope.READ_VEHICLE_INFO,
    Scope.READ_VIN,
]

CONFIGURABLE_SCOPES = [scope for scope in Scope if scope not in REQUIRED_SCOPES]

DEFAULT_SCOPES = [
    Scope.READ_BATTERY,
    Scope.READ_CHARGE,
    Scope.READ_LOCATION,
    Scope.READ_ODOMETER,
    Scope.READ_SECURITY,
    Scope.READ_VEHICLE_INFO,
    Scope.READ_VIN,
    Scope.CONTROL_CHARGE,
]
