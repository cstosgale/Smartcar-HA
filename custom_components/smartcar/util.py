from collections.abc import Mapping
from typing import Any


def unique_id_from_entry_data(data: dict) -> str:
    return " ".join(sorted(data["vehicles"].keys())).lower()


def vins_from_entry_data(data: Mapping[str, Any]) -> str:
    return " ".join(sorted([vehicle["vin"] for vehicle in data["vehicles"].values()]))
