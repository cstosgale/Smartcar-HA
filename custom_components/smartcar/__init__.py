# custom_components/smartcar/__init__.py

import asyncio
import logging

from aiohttp import ClientResponseError
from homeassistant.exceptions import ConfigEntryAuthFailed

from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant
from homeassistant.helpers import config_entry_oauth2_flow, device_registry as dr
from homeassistant.helpers.update_coordinator import UpdateFailed

from .coordinator import SmartcarVehicleCoordinator
from .const import DOMAIN, PLATFORMS, API_BASE_URL_V2

_LOGGER = logging.getLogger(__name__)


async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Set up Smartcar from a config entry."""
    hass.data.setdefault(DOMAIN, {})
    implementation = (
        await config_entry_oauth2_flow.async_get_config_entry_implementation(
            hass, entry
        )
    )
    session = config_entry_oauth2_flow.OAuth2Session(hass, entry, implementation)

    hass.data[DOMAIN][entry.entry_id] = {"session": session, "coordinators": {}}
    entry_data = hass.data[DOMAIN][entry.entry_id]

    try:
        _LOGGER.info("Fetching Smartcar vehicle IDs...")
        vehicle_list_resp = await session.async_request(
            "get", f"{API_BASE_URL_V2}/vehicles"
        )
        vehicle_list_resp.raise_for_status()
        vehicle_list_data = await vehicle_list_resp.json()
        vehicle_ids = vehicle_list_data.get("vehicles", [])
        _LOGGER.info("Found %d vehicle IDs", len(vehicle_ids))
    except ClientResponseError as err:
        if err.status in (401, 403):
            raise ConfigEntryAuthFailed(
                f"Auth error fetching vehicle list: {err.status}"
            ) from err
        else:
            _LOGGER.exception("HTTP Error fetching vehicle list")
            return False
    except ConfigEntryAuthFailed:
        raise  # Already logged by helper potentially
    except Exception:
        _LOGGER.exception("Unexpected error fetching vehicle list")
        return False

    if not vehicle_ids:
        _LOGGER.warning("No vehicle IDs found.")
        return True

    device_registry = dr.async_get(hass)
    setup_tasks = [
        async_setup_single_vehicle(hass, entry, session, vid, device_registry)
        for vid in vehicle_ids
    ]
    results = await asyncio.gather(*setup_tasks, return_exceptions=True)

    auth_failed = any(
        isinstance(res, ConfigEntryAuthFailed)
        for res in results
        if isinstance(res, Exception)
    )
    any_failed = any(isinstance(res, Exception) for res in results)

    if auth_failed:
        _LOGGER.error("Authentication failed during setup of at least one vehicle.")
        return False
    if any_failed:
        _LOGGER.warning("One or more vehicles failed non-critical setup steps.")
    if not entry_data["coordinators"]:
        _LOGGER.warning("No vehicles were successfully set up.")
        return True

    # Log stored scopes once on successful setup
    stored_token_info = entry.data.get("token")
    if stored_token_info:
        _LOGGER.info("Using token with scopes: %s", stored_token_info.get("scope"))
    else:
        _LOGGER.warning("No token information found in ConfigEntry data!")

    _LOGGER.debug("Forwarding setup to platforms: %s", PLATFORMS)
    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
    return True


# --- Corrected Function Signature Below ---
async def async_setup_single_vehicle(
    hass: HomeAssistant,
    entry: ConfigEntry,
    session: config_entry_oauth2_flow.OAuth2Session,
    vehicle_id: str,
    device_registry: dr.DeviceRegistry,
) -> None:
    # --- End Corrected Signature ---
    """Set up a single vehicle, register device, create coordinator. Raises exceptions on failure."""
    vin = None
    entry_data = hass.data[DOMAIN][entry.entry_id]
    try:
        # Get VIN (read_vin scope)
        _LOGGER.debug("Fetching VIN for vehicle ID: %s", vehicle_id)
        vin_resp = await session.async_request(
            "get", f"{API_BASE_URL_V2}/vehicles/{vehicle_id}/vin"
        )
        vin_resp.raise_for_status()
        vin_data = await vin_resp.json()
        vin = vin_data.get("vin")
        if not vin:
            raise ValueError("Missing VIN")

        # Get Attributes (read_vehicle_info scope)
        _LOGGER.debug("Fetching attributes for VIN: %s (ID: %s)", vin, vehicle_id)
        attr_resp = await session.async_request(
            "get", f"{API_BASE_URL_V2}/vehicles/{vehicle_id}"
        )
        attr_resp.raise_for_status()
        vehicle_info = await attr_resp.json()
        make = vehicle_info.get("make")
        model = vehicle_info.get("model")
        year = vehicle_info.get("year")

        # Register device
        device_registry.async_get_or_create(
            config_entry_id=entry.entry_id,
            identifiers={(DOMAIN, vin)},
            manufacturer=make,
            model=f"{model} ({year})" if model and year else model,
            name=f"{make} {model}" if make and model else f"Smartcar {vin[-4:]}",
        )
        _LOGGER.info("Registered device for VIN: %s", vin)

        # Create and Store Coordinator
        coordinator = SmartcarVehicleCoordinator(hass, session, vehicle_id, vin, entry)
        await coordinator.async_config_entry_first_refresh()
        entry_data["coordinators"][vin] = coordinator
        _LOGGER.debug("Coordinator created and initial data fetched for VIN: %s", vin)

    except ClientResponseError as err:
        if err.status in (401, 403):
            raise ConfigEntryAuthFailed(
                f"Auth error [{err.status}] during vehicle setup"
            ) from err
        else:
            raise UpdateFailed(
                f"API error during setup: {err.status}"
            ) from err  # Raise UpdateFailed for non-auth HTTP errors
    except ConfigEntryAuthFailed:
        raise  # Propagate auth failure
    except Exception as err:
        raise err  # Propagate other errors


async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Unload a config entry."""
    _LOGGER.info("Unloading Smartcar entry %s", entry.entry_id)
    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
    if unload_ok:
        hass.data[DOMAIN].pop(entry.entry_id, None)
        _LOGGER.info("Unloaded Smartcar data for entry %s", entry.entry_id)
    return unload_ok
