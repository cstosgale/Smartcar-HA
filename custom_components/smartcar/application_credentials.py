# custom_components/smartcar/application_credentials.py
# --- Using Withings Pattern ---

from homeassistant.core import HomeAssistant
from homeassistant.components.application_credentials import (
    AuthorizationServer,
    ClientCredential,
)
from homeassistant.helpers import config_entry_oauth2_flow
from homeassistant.helpers.config_entry_oauth2_flow import AbstractOAuth2Implementation

from .const import OAUTH2_AUTHORIZE, OAUTH2_TOKEN


async def async_get_authorization_server(hass: HomeAssistant) -> AuthorizationServer:
    """Return authorization server details."""
    return AuthorizationServer(
        authorize_url=OAUTH2_AUTHORIZE,
        token_url=OAUTH2_TOKEN,
    )


async def async_get_auth_implementation(
    hass: HomeAssistant,
    auth_domain: str,
    credential: ClientCredential,
) -> AbstractOAuth2Implementation:
    """Return auth implementation for Smartcar."""
    return config_entry_oauth2_flow.LocalOAuth2Implementation(
        hass=hass,
        domain=auth_domain,
        client_id=credential.client_id,
        client_secret=credential.client_secret,
        authorize_url=OAUTH2_AUTHORIZE,
        token_url=OAUTH2_TOKEN,
    )
