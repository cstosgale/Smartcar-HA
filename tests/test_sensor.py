"""Test sensors."""

from dataclasses import dataclass
import datetime as dt
from typing import Any, cast
from unittest.mock import AsyncMock

from homeassistant.core import HomeAssistant, State
from homeassistant.helpers import device_registry as dr, entity_registry as er
from homeassistant.helpers.entity_component import async_update_entity
from homeassistant.util.dt import utcnow
import pytest
from pytest_homeassistant_custom_component.common import (
    MockConfigEntry,
    async_fire_time_changed,
    mock_restore_cache_with_extra_data,
)
from pytest_homeassistant_custom_component.test_util.aiohttp import AiohttpClientMocker
from syrupy.assertion import SnapshotAssertion

from custom_components.smartcar.const import (
    DEFAULT_ENABLED_ENTITY_DESCRIPTION_KEYS,
    REQUIRED_SCOPES,
    EntityDescriptionKey,
)
from custom_components.smartcar.entity import SmartcarEntity
from custom_components.smartcar.sensor import SmartcarSensorDescription

from . import setup_added_integration, setup_integration


@pytest.mark.usefixtures("enable_all_entities")
@pytest.mark.parametrize("vehicle_fixture", ["vw_id_4"])
async def test_polling_updates(
    hass: HomeAssistant,
    mock_config_entry: MockConfigEntry,
    aioclient_mock: AiohttpClientMocker,
    snapshot: SnapshotAssertion,
    vehicle: AsyncMock,
    vehicle_fixture: str,
    vehicle_attributes: dict,
    device_registry: dr.DeviceRegistry,
    entity_registry: er.EntityRegistry,
) -> None:
    """Test sensors when polling is enabled."""

    entity_id = "sensor.vw_id_4_odometer"
    exepcted_calls = 0

    await setup_integration(hass, mock_config_entry)
    assert entity_registry.async_get(entity_id) == snapshot(
        name="sensor.vw_id_4_odometer-registry"
    )
    assert hass.states.get(entity_id) == snapshot(name=entity_id)

    # check the first refresh based update (a full batch)
    batch_request_mock_call = aioclient_mock.mock_calls[-1]
    assert aioclient_mock.call_count == (exepcted_calls := exepcted_calls + 1)
    assert batch_request_mock_call == snapshot(
        name=f"{entity_id}-api-full-batch-request"
    )

    # trigger a polling based update (another full batch update)
    async_fire_time_changed(hass, utcnow() + dt.timedelta(hours=6))
    await hass.async_block_till_done()
    assert aioclient_mock.call_count == (exepcted_calls := exepcted_calls + 1)
    assert aioclient_mock.mock_calls[-1] == batch_request_mock_call

    # trigger an update for a single entity (the batch update should only
    # include a request for the odometer)
    await async_update_entity(hass, entity_id)
    assert aioclient_mock.call_count == (exepcted_calls := exepcted_calls + 1)
    assert aioclient_mock.mock_calls[-1] == snapshot(
        name=f"{entity_id}-api-odometer-request"
    )

    # disable the entity & update it -- this should not result in any api calls
    entity_registry.async_update_entity(
        entity_id, disabled_by=er.RegistryEntryDisabler.USER
    )
    await async_update_entity(hass, entity_id)
    assert aioclient_mock.call_count == exepcted_calls  # no update should have occurred

    # trigger a polling based update (which should not include the disabled
    # odometer in the request)
    async_fire_time_changed(hass, utcnow() + dt.timedelta(hours=12))
    await hass.async_block_till_done()
    batch_request_without_odometer_mock_call = aioclient_mock.mock_calls[-1]
    assert aioclient_mock.call_count == (exepcted_calls := exepcted_calls + 1)
    assert batch_request_without_odometer_mock_call == snapshot(
        name=f"{entity_id}-api-batch-without-odometer-request"
    )


@pytest.mark.usefixtures("enable_all_entities")
@pytest.mark.parametrize("vehicle_fixture", ["vw_id_4"])
async def test_update_with_polling_disabled(
    hass: HomeAssistant,
    mock_config_entry: MockConfigEntry,
    aioclient_mock: AiohttpClientMocker,
    snapshot: SnapshotAssertion,
    vehicle: AsyncMock,
    device_registry: dr.DeviceRegistry,
    entity_registry: er.EntityRegistry,
) -> None:
    """Test sensors when polling is disabled."""

    mock_config_entry.add_to_hass(hass)
    hass.config_entries.async_update_entry(
        mock_config_entry,
        pref_disable_polling=True,
    )

    entity_id = "sensor.vw_id_4_odometer"
    exepcted_calls = 0

    await setup_added_integration(hass, mock_config_entry)

    assert entity_registry.async_get(entity_id) == snapshot(
        name="sensor.vw_id_4_odometer-registry"
    )
    assert hass.states.get(entity_id) == snapshot(name=entity_id)

    # no requests should have been made during setup when polling is disabled.
    # this is different from the default behavior, but intended to help reduce
    # excessive api usage.
    assert aioclient_mock.call_count == exepcted_calls

    await async_update_entity(hass, entity_id)

    assert aioclient_mock.call_count == (exepcted_calls := exepcted_calls + 1)
    assert aioclient_mock.mock_calls[-1] == snapshot(
        name=f"{entity_id}-api-odometer-request"
    )

    entity_registry.async_update_entity(
        entity_id, disabled_by=er.RegistryEntryDisabler.USER
    )

    # after another update, no more api calls should have occurred
    await async_update_entity(hass, entity_id)
    assert aioclient_mock.call_count == exepcted_calls


@pytest.mark.parametrize("vehicle_fixture", ["vw_id_4"])
@pytest.mark.parametrize("enabled_scopes", [REQUIRED_SCOPES + ["read_battery"]])
@pytest.mark.parametrize(
    "enabled_entities",
    [DEFAULT_ENABLED_ENTITY_DESCRIPTION_KEYS | {EntityDescriptionKey.BATTERY_CAPACITY}],
)
async def test_limited_scopes(
    hass: HomeAssistant,
    mock_config_entry: MockConfigEntry,
    aioclient_mock: AiohttpClientMocker,
    snapshot: SnapshotAssertion,
    vehicle: AsyncMock,
    device_registry: dr.DeviceRegistry,
    entity_registry: er.EntityRegistry,
) -> None:
    """Test sensors when only limited scopes are enabled."""

    entity_id = "sensor.vw_id_4_battery"
    exepcted_calls = 0

    await setup_integration(hass, mock_config_entry)
    assert entity_registry.async_get(entity_id) == snapshot(
        name=f"{entity_id}-registry"
    )
    assert hass.states.get(entity_id) == snapshot(name=entity_id)
    assert aioclient_mock.call_count == (exepcted_calls := exepcted_calls + 1)
    assert aioclient_mock.mock_calls[-1] == snapshot(
        name=f"{entity_id}-api-battery-request"
    )


@pytest.mark.usefixtures("enable_all_entities")
@pytest.mark.parametrize("vehicle_fixture", ["vw_id_4"])
@pytest.mark.parametrize("api_respone_type", ["imperial"])
async def test_unit_conversion(
    hass: HomeAssistant,
    mock_config_entry: MockConfigEntry,
    aioclient_mock: AiohttpClientMocker,
    snapshot: SnapshotAssertion,
    vehicle: AsyncMock,
    device_registry: dr.DeviceRegistry,
    entity_registry: er.EntityRegistry,
) -> None:
    """Test sensors that could return imperial units."""

    entity_id = "sensor.vw_id_4_odometer"
    exepcted_calls = 0

    await setup_integration(hass, mock_config_entry)
    assert entity_registry.async_get(entity_id) == snapshot(
        name=f"{entity_id}-registry"
    )
    assert hass.states.get(entity_id) == snapshot(name=entity_id)

    assert aioclient_mock.call_count == (exepcted_calls := exepcted_calls + 1)
    assert aioclient_mock.mock_calls[-1] == snapshot(
        name=f"{entity_id}-api-full-batch-request"
    )

    await async_update_entity(hass, entity_id)

    assert aioclient_mock.call_count == (exepcted_calls := exepcted_calls + 1)
    assert aioclient_mock.mock_calls[-1] == snapshot(
        name=f"{entity_id}-api-odometer-request"
    )


@pytest.mark.parametrize(
    (
        "entity_id",
        "stored_data",
        "expected_state",
        "expected_data",
    ),
    [
        (
            "sensor.vw_id_4_battery",
            {"raw_value": 0.34},
            "34",
            {"battery": {"percentRemaining": 0.34}},
        ),
        (
            "sensor.vw_id_4_range",
            {"raw_value": 45.2, "unit_system": "imperial"},
            "72.7423488",
            {"battery": {"range": 45.2}, "battery:unit_system": "imperial"},
        ),
    ],
    ids=["value_only", "value_and_unit_system"],
)
@pytest.mark.parametrize("vehicle_fixture", ["vw_id_4"])
async def test_restore_state(
    hass: HomeAssistant,
    mock_config_entry: MockConfigEntry,
    vehicle_attributes: dict,
    entity_id: str,
    stored_data: dict,
    expected_state: Any,
    expected_data: dict,
) -> None:
    """Test sensor restore state."""

    mock_restore_cache_with_extra_data(
        hass,
        (
            (
                State(
                    entity_id,
                    "does-not-matter-for-this-test",
                ),
                stored_data,
            ),
        ),
    )

    # await setup_integration(hass, mock_config_entry)
    mock_config_entry.add_to_hass(hass)
    hass.config_entries.async_update_entry(
        mock_config_entry,
        pref_disable_polling=True,
    )

    await setup_added_integration(hass, mock_config_entry)

    coordinator = mock_config_entry.runtime_data.coordinators[vehicle_attributes["vin"]]
    state = hass.states.get(entity_id)
    assert state
    assert state.state == expected_state
    assert coordinator.data == expected_data


async def test_async_update_internals(
    hass: HomeAssistant,
) -> None:
    """Test that coordinator does not have sensor added during update of a disabled entity."""

    @dataclass(kw_only=True)
    class MockCoordinator:
        vin: str | None = None

    @dataclass(kw_only=True)
    class MockEntityDescription:
        key: str | None = None

    @dataclass(kw_only=True)
    class MockRegistryEntry:
        disabled: bool = False

    coordinator = MockCoordinator()
    description = MockEntityDescription()
    entity = SmartcarEntity(cast(Any, coordinator), description)
    entity.registry_entry = MockRegistryEntry()

    # prove that the failure occurs when enabled
    with pytest.raises(AttributeError) as excinfo:
        await entity.async_update()
    assert excinfo.value.obj == coordinator
    assert excinfo.value.name == "batch_sensor"

    # and then it does not when it's disabled
    entity.registry_entry.disabled = True
    await entity.async_update()


def test_entity_registry_enabled_default_readonly(
    hass: HomeAssistant,
) -> None:
    """Test entity_registry_enabled_default is readonly."""

    with pytest.raises(AttributeError) as excinfo:
        SmartcarSensorDescription(
            key="mock_description",
            value_key_path="mock_path",
            entity_registry_enabled_default=False,
        )

    assert (
        excinfo.value.args[0]
        == "readonly; configure via smartcar.const.DEFAULT_ENABLED_ENTITY_DESCRIPTION_KEYS"
    )
