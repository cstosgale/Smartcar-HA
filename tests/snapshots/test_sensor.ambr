# serializer version: 1
# name: test_limited_scopes[enabled_entities0-enabled_scopes0-vw_id_4][sensor.vw_id_4_battery-api-battery-request]
  tuple(
    'post',
    URL('http://test.local/v2.0/vehicles/a1d50709-3502-4faa-ba43-a5c7565e6a09/batch'),
    dict({
      'requests': list([
        dict({
          'path': '/battery',
        }),
      ]),
    }),
    dict({
      'authorization': 'Bearer mock-token',
    }),
  )
# ---
# name: test_limited_scopes[enabled_entities0-enabled_scopes0-vw_id_4][sensor.vw_id_4_battery-registry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': dict({
      'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
    }),
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'sensor',
    'entity_category': None,
    'entity_id': 'sensor.vw_id_4_battery',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': <SensorDeviceClass.BATTERY: 'battery'>,
    'original_icon': None,
    'original_name': 'Battery',
    'platform': 'smartcar',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': None,
    'unique_id': 'VIWP1AB29P15LA85784N_battery_level',
    'unit_of_measurement': '%',
  })
# ---
# name: test_limited_scopes[enabled_entities0-enabled_scopes0-vw_id_4][sensor.vw_id_4_battery]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'age': '2025-05-09T15:40:57+00:00',
      'device_class': 'battery',
      'fetched_at': '2025-05-09T17:27:26.712000+00:00',
      'friendly_name': 'VW ID.4 Battery',
      'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
      'unit_of_measurement': '%',
    }),
    'context': <ANY>,
    'entity_id': 'sensor.vw_id_4_battery',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '68',
  })
# ---
# name: test_polling_updates[vw_id_4][sensor.vw_id_4_odometer-api-batch-without-odometer-request]
  tuple(
    'post',
    URL('http://test.local/v2.0/vehicles/a1d50709-3502-4faa-ba43-a5c7565e6a09/batch'),
    dict({
      'requests': list([
        dict({
          'path': '/battery',
        }),
        dict({
          'path': '/battery/nominal_capacity',
        }),
        dict({
          'path': '/charge',
        }),
        dict({
          'path': '/charge/limit',
        }),
        dict({
          'path': '/engine/oil',
        }),
        dict({
          'path': '/fuel',
        }),
        dict({
          'path': '/location',
        }),
        dict({
          'path': '/security',
        }),
        dict({
          'path': '/tires/pressure',
        }),
      ]),
    }),
    dict({
      'authorization': 'Bearer mock-token',
    }),
  )
# ---
# name: test_polling_updates[vw_id_4][sensor.vw_id_4_odometer-api-full-batch-request]
  tuple(
    'post',
    URL('http://test.local/v2.0/vehicles/a1d50709-3502-4faa-ba43-a5c7565e6a09/batch'),
    dict({
      'requests': list([
        dict({
          'path': '/battery',
        }),
        dict({
          'path': '/battery/nominal_capacity',
        }),
        dict({
          'path': '/charge',
        }),
        dict({
          'path': '/charge/limit',
        }),
        dict({
          'path': '/engine/oil',
        }),
        dict({
          'path': '/fuel',
        }),
        dict({
          'path': '/location',
        }),
        dict({
          'path': '/odometer',
        }),
        dict({
          'path': '/security',
        }),
        dict({
          'path': '/tires/pressure',
        }),
      ]),
    }),
    dict({
      'authorization': 'Bearer mock-token',
    }),
  )
# ---
# name: test_polling_updates[vw_id_4][sensor.vw_id_4_odometer-api-odometer-request]
  tuple(
    'post',
    URL('http://test.local/v2.0/vehicles/a1d50709-3502-4faa-ba43-a5c7565e6a09/batch'),
    dict({
      'requests': list([
        dict({
          'path': '/odometer',
        }),
      ]),
    }),
    dict({
      'authorization': 'Bearer mock-token',
    }),
  )
# ---
# name: test_polling_updates[vw_id_4][sensor.vw_id_4_odometer-registry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': dict({
      'state_class': <SensorStateClass.TOTAL_INCREASING: 'total_increasing'>,
    }),
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'sensor',
    'entity_category': None,
    'entity_id': 'sensor.vw_id_4_odometer',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
      'sensor': dict({
        'suggested_display_precision': 2,
      }),
    }),
    'original_device_class': <SensorDeviceClass.DISTANCE: 'distance'>,
    'original_icon': None,
    'original_name': 'Odometer',
    'platform': 'smartcar',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': None,
    'unique_id': 'VIWP1AB29P15LA85784N_odometer',
    'unit_of_measurement': <UnitOfLength.KILOMETERS: 'km'>,
  })
# ---
# name: test_polling_updates[vw_id_4][sensor.vw_id_4_odometer]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'age': '2025-05-09T15:40:52+00:00',
      'device_class': 'distance',
      'fetched_at': '2025-05-09T17:27:25.401000+00:00',
      'friendly_name': 'VW ID.4 Odometer',
      'state_class': <SensorStateClass.TOTAL_INCREASING: 'total_increasing'>,
      'unit_of_measurement': <UnitOfLength.KILOMETERS: 'km'>,
    }),
    'context': <ANY>,
    'entity_id': 'sensor.vw_id_4_odometer',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '21919',
  })
# ---
# name: test_restore_sensor_save_state[vw_id_4-value_and_timestampes]
  list([
    dict({
      'data_age': '2025-05-29T19:47:32+00:00',
      'fetched_at': '2025-05-29T20:09:57+00:00',
      'raw_value': 45.2,
    }),
  ])
# ---
# name: test_restore_sensor_save_state[vw_id_4-value_and_unit_system]
  list([
    dict({
      'raw_value': 45.2,
      'unit_system': 'imperial',
    }),
  ])
# ---
# name: test_restore_sensor_save_state[vw_id_4-value_only]
  list([
    dict({
      'raw_value': 0.34,
    }),
  ])
# ---
# name: test_unit_conversion[imperial-vw_id_4][sensor.vw_id_4_odometer-api-full-batch-request]
  tuple(
    'post',
    URL('http://test.local/v2.0/vehicles/a1d50709-3502-4faa-ba43-a5c7565e6a09/batch'),
    dict({
      'requests': list([
        dict({
          'path': '/battery',
        }),
        dict({
          'path': '/battery/nominal_capacity',
        }),
        dict({
          'path': '/charge',
        }),
        dict({
          'path': '/charge/limit',
        }),
        dict({
          'path': '/engine/oil',
        }),
        dict({
          'path': '/fuel',
        }),
        dict({
          'path': '/location',
        }),
        dict({
          'path': '/odometer',
        }),
        dict({
          'path': '/security',
        }),
        dict({
          'path': '/tires/pressure',
        }),
      ]),
    }),
    dict({
      'authorization': 'Bearer mock-token',
    }),
  )
# ---
# name: test_unit_conversion[imperial-vw_id_4][sensor.vw_id_4_odometer-api-odometer-request]
  tuple(
    'post',
    URL('http://test.local/v2.0/vehicles/a1d50709-3502-4faa-ba43-a5c7565e6a09/batch'),
    dict({
      'requests': list([
        dict({
          'path': '/odometer',
        }),
      ]),
    }),
    dict({
      'authorization': 'Bearer mock-token',
    }),
  )
# ---
# name: test_unit_conversion[imperial-vw_id_4][sensor.vw_id_4_odometer-registry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': dict({
      'state_class': <SensorStateClass.TOTAL_INCREASING: 'total_increasing'>,
    }),
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'sensor',
    'entity_category': None,
    'entity_id': 'sensor.vw_id_4_odometer',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
      'sensor': dict({
        'suggested_display_precision': 2,
      }),
    }),
    'original_device_class': <SensorDeviceClass.DISTANCE: 'distance'>,
    'original_icon': None,
    'original_name': 'Odometer',
    'platform': 'smartcar',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': None,
    'unique_id': 'VIWP1AB29P15LA85784N_odometer',
    'unit_of_measurement': <UnitOfLength.KILOMETERS: 'km'>,
  })
# ---
# name: test_unit_conversion[imperial-vw_id_4][sensor.vw_id_4_odometer]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'age': '2019-10-24T00:43:46+00:00',
      'device_class': 'distance',
      'friendly_name': 'VW ID.4 Odometer',
      'state_class': <SensorStateClass.TOTAL_INCREASING: 'total_increasing'>,
      'unit_of_measurement': <UnitOfLength.KILOMETERS: 'km'>,
    }),
    'context': <ANY>,
    'entity_id': 'sensor.vw_id_4_odometer',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '60879.874176',
  })
# ---
# name: test_update_with_polling_disabled[vw_id_4][sensor.vw_id_4_odometer-api-odometer-request]
  tuple(
    'post',
    URL('http://test.local/v2.0/vehicles/a1d50709-3502-4faa-ba43-a5c7565e6a09/batch'),
    dict({
      'requests': list([
        dict({
          'path': '/odometer',
        }),
      ]),
    }),
    dict({
      'authorization': 'Bearer mock-token',
    }),
  )
# ---
# name: test_update_with_polling_disabled[vw_id_4][sensor.vw_id_4_odometer-registry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': dict({
      'state_class': <SensorStateClass.TOTAL_INCREASING: 'total_increasing'>,
    }),
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'sensor',
    'entity_category': None,
    'entity_id': 'sensor.vw_id_4_odometer',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
      'sensor': dict({
        'suggested_display_precision': 2,
      }),
    }),
    'original_device_class': <SensorDeviceClass.DISTANCE: 'distance'>,
    'original_icon': None,
    'original_name': 'Odometer',
    'platform': 'smartcar',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': None,
    'unique_id': 'VIWP1AB29P15LA85784N_odometer',
    'unit_of_measurement': <UnitOfLength.KILOMETERS: 'km'>,
  })
# ---
# name: test_update_with_polling_disabled[vw_id_4][sensor.vw_id_4_odometer]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'device_class': 'distance',
      'friendly_name': 'VW ID.4 Odometer',
      'state_class': <SensorStateClass.TOTAL_INCREASING: 'total_increasing'>,
      'unit_of_measurement': <UnitOfLength.KILOMETERS: 'km'>,
    }),
    'context': <ANY>,
    'entity_id': 'sensor.vw_id_4_odometer',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'unavailable',
  })
# ---
